src/ConnectedMode/Binding/ServerConnectionConfigurationProvider.cs

```diff
@@ -0,0 +1,109 @@
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Generic;
+using System.ComponentModel.Composition;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using SonarLint.VisualStudio.ConnectedMode.Persistence;
+using SonarLint.VisualStudio.Core;
+using SonarLint.VisualStudio.Core.Binding;
+using SonarLint.VisualStudio.Infrastructure.VS;
+using SonarLint.VisualStudio.SLCore.Common.Helpers;
+using SonarLint.VisualStudio.SLCore.Service.Connection.Models;
+
+namespace SonarLint.VisualStudio.ConnectedMode.Binding
+{
+    [Export(typeof(IServerConnectionConfigurationProvider))]
+    [PartCreationPolicy(CreationPolicy.Shared)]
+    internal class ServerConnectionConfigurationProvider : IServerConnectionConfigurationProvider
+    {
+        private readonly ISolutionBindingRepository solutionBindingRepository;
+        private readonly IThreadHandling threadHandling;
+        private readonly IConnectionIdHelper connectionIdHelper;
+
+        [ImportingConstructor]
+        [ExcludeFromCodeCoverage]
+        public ServerConnectionConfigurationProvider(ISolutionBindingRepository solutionBindingRepository)
+            : this(solutionBindingRepository, ThreadHandling.Instance, new ConnectionIdHelper())
+        {
+        }
+
+        internal ServerConnectionConfigurationProvider(ISolutionBindingRepository solutionBindingRepository, IThreadHandling threadHandling, IConnectionIdHelper connectionIdHelper)
+        {
+            this.solutionBindingRepository = solutionBindingRepository;
+            this.threadHandling = threadHandling;
+            this.connectionIdHelper = connectionIdHelper;
+        }
+
+        public IEnumerable<T> GetServerConnectionConfiguration<T>() where T : ServerConnectionConfiguration
+        {
+            threadHandling.ThrowIfOnUIThread();
+
+            var bindingList = new List<ServerConnectionConfiguration>();
+
+            var bindings = solutionBindingRepository.List().Distinct(new BoundSonarQubeProjectUriComparer());
+
+            foreach (var binding in bindings)
+            {
+                var connectionID = connectionIdHelper.GetConnectionIdFromUri(binding.ServerUri, binding.Organization?.Key);
+
+                if (binding.ServerUri == ConnectionIdHelper.SonarCloudUri)
+                {
+                    bindingList.Add(new SonarCloudConnectionConfigurationDto(connectionID, true, binding.Organization?.Key));
+                }
+                else
+                {
+                    bindingList.Add(new SonarQubeConnectionConfigurationDto(connectionID, true, binding.ServerUri.ToString()));
+                }
+            }
+
+            return bindingList.OfType<T>();
```

Comment:
This doesn't make much sense after you removed caching. I would return the list of ServerConnectionConfiguration and let the caller filter that instead

Reply:
caller can still do it if they pass ServerConnectionConfiguration as T.

Reply:
But if they don't, they will do IO twice. There's no need for this generic now.